Postgres database

kavitharaju opened the pull request:
- Adds new vector database implementation, Postgres-with-pgvector, along with existing Chroma DB #17 
- Implements handling of multiple search domains, with filtering based on `label`.
- Includes one more demo UI, which uses the new tech stack with postgres DB and openai embeddings.
- The changes have been deployed at https://assistant.bible/

![Screenshot from 2023-07-20 11-44-21](https://github.com/BibleNLP/assistant.bible/assets/18255929/749e4027-3271-434a-a41b-53d66eb7ca87)
The above demo shows how search domain can be selected so that the bot will anwer only from those set of documents.

woodwardmw commented on app/custom_exceptions.py:
```diff
@@ -40,6 +40,14 @@ def __init__(self, detail):
+    '''Format for errors from ChromaDB's APIs'''
+    def __init__(self, detail):
+        super().__init__()
+        self.name = "Error from Postgres Databse"
```
Small typo here - "Database"

kavitharaju replied:
Fixing this

woodwardmw reviewed (approved):
This looks good to me! One thing  in trying out the deployed version, it seems the sources displayed are always the first three sources in the collection, and not necessarily those that were the closest match to the query.

kavitharaju commented:
Oh, I think you are right. It only show the first 3 in postgres, but chroma does it correctly. Like you said before, once the implementation is done, we will have to do a few rounds of testings.. and will have to see which kind of vector search algos work properly and all... Now in postgres I use the `<=>` operator in postgres, out of the 3 they have
https://github.com/kavitharaju/assistant.bible/blob/68c1341bc2cc008946e12de1d25470625b92181c/app/core/vectordb/postgres4langchain.py#L115

kavitharaju commented:
The problem of, all queries being answered with first 3 entries, should be solved now.
I had missed generating the embedding for text before insertion. So all comparison were being done against null vectors :grimacing:

kavitharaju commented:
The merge conflicts are resolved. 
But the linting errors from the pulled code are making the ci-cd to fail.

kavitharaju commented:
- Ran into a few minor issues. Fixed those encountered
- Separated out auth decorator for chat and other data managment APIs #18 
    - The data upload APIs will require an admin token which can be set as environment variable in a deployment
    - The chat websocket could recieve a user token from supabase and determine if he has access on requested sources(this is being implemented separately by @woodwardmw) 
- Updated deployment at assistant.bible with all the changes up until now
   - Demo 1 has Whisper and uses chroma db
   - Demo 2 has only text input. Uses Postgres and allows filtering source.
- Added full Gen 1, instead of first 5 verses.
